File: battleship.py
import random

# Represents a ship object on the board
class Ship:
	def __init__(self, name, length, isHorizontal):
		self.name = name
		self.length = length

		# True if the ship is horizontally oriented
		self.isHorizontal = isHorizontal 

		# Coordinates are only set when the ship is placed successfully
		self.x = None # Leftmost x-coordinate
		self.y = None # Topmost y-coordinate

		# True if the ship has been sunk
		self.isSunk = False 

# Represents one square of the board
class GridSquare:
	def __init__(self, x, y):
		self.x = x
		self.y = y

		# Holds a reference to the Ship object occupying this square, if any
		self.ship = None 

		# True if the opponent has guessed this square
		self.isGuessed = False
		# TODO: change this to use a better state representation

	# Returns a single-character representation of the square
	# Key:
	# 	O = missed
	# 	X = hit
	# 	_ = not guessed
	# 	* = contains ship (only visible if can_see_ships is True)
	def gs_string(self, can_see_ships):
		if self.isGuessed and self.ship is not None:
			return 'X'
		elif self.ship is not None and can_see_ships:
			return '*'
		elif self.isGuessed:
			return 'O'
		else:
			return '_'

# Represents the board
class Grid:
	# If addRandomShips is True, randomly places 5 ships on the board
	def __init__(self, add_random_ships = False):
		self.w = 10 # Default width of 10
		self.h = 10 # Default height of 10

		# Initialize an empty wxh grid array and empty ships array
		self.grid_array = [[GridSquare(x, y) for x in range(self.w)] for y in range(self.h)]
		self.ships = []

		# Randomly place the ships
		if add_random_ships:
			ships_to_add = [
				Ship('Aircraft Carrier', 5, False),
				Ship('Battleship', 4, True),
				Ship('Submarine', 3, False),
				Ship('Cruiser', 3, True),
				Ship('Destroyer', 2, False)
			]
			while len(ships_to_add) > 0:
				rand_x = random.randint(0, 9)
				rand_y = random.randint(0, 9)
				s = ships_to_add[0]
				if self.add_ship(s, rand_x, rand_y):
					ships_to_add.remove(s)

	# Checks if the board is 'ready', e.g. all 5 ships have been added
	def check_if_ready(self):
		return len(self.ships) == 5

	# Adds the given ship to the grid at the given coordinates
	# Returns True if successful, False if invalid placement
	def add_ship(self, ship, x, y):
		# First check to see if this is in bounds of the grid
		if (x < 0 or (ship.isHorizontal and (x + ship.length) >= self.w)
			or y < 0 or (not ship.isHorizontal and (y + ship.length) >= self.h)):
			return False

		# TODO: add a check here to make sure we can add a ship
		# of this length (i.e. can't have more than 2 length-3 ships)

		# Next check to see if this will overlap with any other ships
		curX = x
		curY = y
		for i in range(ship.length):
			if self.grid_array[curX][curY].ship is not None:
				return False
			if ship.isHorizontal: 
				curX += 1
			else: 
				curY += 1

		# Add the ship to the grid
		ship.x = x
		ship.y = y
		for i in range(ship.length):
			self.grid_array[x][y].ship = ship
			if ship.isHorizontal:
				x += 1
			else:
				y += 1
		self.ships.append(ship)
		return True

	# Makes a guess at the given coordinates and updates the board
	# Returns the response message
	def make_guess(self, x, y):
		# First check if the guess in in bounds
		if x < 0 or x >= self.w or y < 0 or y >= self.h:
			return 'Out of bounds.'

		gs = self.grid_array[x][y]

		# Next check if the square has already been guessed
		if gs.isGuessed:
			return 'Already guessed.'

		# Make the guess
		gs.isGuessed = True
		if gs.ship is None:
			return 'Miss!'
		else:
			if self.check_sunk(gs.ship):
				# Check to see if all ships have been sunk
				for ship in self.ships:
					if not self.check_sunk(ship):
						return 'Hit and sink! You sunk the ' + gs.ship.name
				return 'Hit and sink! You sunk all the ships -- you win!'
			else:
				return 'Hit!'

	# Checks if the given ship is fully sunk (i.e. all of its coordinates
	# have been hit)
	# This method does NOT modify the grid or the ship object
	def check_sunk(self, ship):
		curX = ship.x
		curY = ship.y
		if curX is None or curY is None:
			return False

		for i in range(ship.length):
			if not self.grid_array[curX][curY].isGuessed:
				return False
			if ship.isHorizontal:
				curX += 1
			else:
				curY += 1

		return True

	# Prints out the grid in a human readable format
	def print_grid(self, can_see_ships):
		for y in range(self.h):
			row = ''
			for x in range(self.w):
				row += self.grid_array[x][y].gs_string(can_see_ships) + ' '
			print row

File: battleship.js
var invalidGuess = 'Invalid guess';
var alreadyGuessed = 'Already guessed';
var hitShip = 'You hit a ship: ';
var hitAndSunk = 'You hit and sunk a ship: ';
var miss = 'You missed :(';

// Initializes a ship with the given length and orientation
function Ship(name, length, isHorizontal) {
	this.name = name;
	this.length = length;
	this.isHorizontal = isHorizontal; 
	this.isSunk = false; // True when the opponent sinks this ship

	// The leftmost and topmost coordinates of this ship
	// These values are only set when the ship is successfully placed on the grid
	this.x = null; // Leftmost x-coordinate of this ship
	this.y = null; // Topmost y-coordinate of this ship
}

// Initializes an empty grid square at the given location
function GridSquare(x, y) {
	this.x = x;
	this.y = y;
	this.ship = null; // True if there is a ship in this square
	this.isGuessed = false; // True if the opponent has 'guessed' this square
}

// Initializes an empty grid of GridSquare objects
// Position (0, 0) is the top left corner of the grid
function Grid() {
	this.w = 10; // Grid width
	this.h = 10; // Grid height

	this.gridArray = [[]];
	for (var x = 0; x < this.w; x++) {
		this.gridArray[x] = [];
		for (var y = 0; y < this.h; y++) {
			var gs = new GridSquare(x, y);
			this.gridArray[x][y] = gs;
		}
	}
}

// Prints out a representation of the grid to the console in a readable format
Grid.prototype.printGrid = function() {
	for (var j = 0; j < this.h; j++) {
		var rowString = j + ': ';
		for (var i = 0; i < this.w; i++) {
			var gsString = ''
			if (this.gridArray[i][j].ship != null) gsString += 's';
			if (this.gridArray[i][j].isGuessed) gsString += 'X';

			if (gsString.length == 0) gsString = '_';
			rowString += gsString + ' ';
		}
		console.log(rowString);
	}
	console.log('\n');
}

// Adds a boat at the specified location with the given orientation
// Returns false if the ship is trying to be placed in an invalid location
// Returns true if the ship was successfully placed
Grid.prototype.addShip = function(ship, x, y) {
	// Make sure that this is 'legal' placement (not off the edges
	// and not conflicting with other ships)
	if (x < 0 || (ship.isHorizontal && (x+ship.length) >= this.w) || 
		y < 0 || (!ship.isHorizontal && (y+ship.length) >= this.h)) {
		return false;
	}

	var curX = x;
	var curY = y;
	for (var i = 0; i < ship.length; i++) {
		if (this.gridArray[curX][curY].ship != null) return false;
		if (ship.isHorizontal) curX += 1;
		else curY += 1;
	}
	
	// Update the appropriate elements in the grid array
	for (var i = 0; i < ship.length; i++) {
		// Set the ship coordinates if necessary
		if (ship.x == null) ship.x = x;
		if (ship.y == null) ship.y = y;

		// Mark the coordinates in the grid array as containing this ship
		this.gridArray[x][y].ship = ship;
		if (ship.isHorizontal) x += 1;
		else y += 1;
	}
	return true;
}

// Helper method to check if the given ship is sunk 
// (i.e. if all of the ship's coordinates have been guessed)
// This method does NOT make any modifications to the grid or ship
// and assumes that the ship has been placed at a valid location
Grid.prototype.checkIfSunk = function(ship) {
	if (ship.x == null || ship.y == null) return false;

	var curX = ship.x;
	var curY = ship.y;
	for (var i = 0; i < ship.length; i++) {
		if (!this.gridArray[curX][curY].isGuessed) return false;
		if (ship.isHorizontal) curX += 1;
		else curY += 1;
	}
	return true;
}

// Makes a guess at the specified location in the grid
// Returns a message to the player
Grid.prototype.makeGuess = function(x, y) {
	// Make sure this is a legal guess
	if (x < 0 || x >= this.w || y < 0 || y >= this.h) return invalidGuess;
	if (this.gridArray[x][y].isGuessed) return alreadyGuessed;

	// First mark this square as guessed
	this.gridArray[x][y].isGuessed = true;

	// Check to see if a ship has been hit
	if (this.gridArray[x][y].ship != null) {
		var ship = this.gridArray[x][y].ship;
		// Check to see if a ship has been sunk
		if (this.checkIfSunk(ship)) return hitAndSunk + ship.name;
		else return hitShip + ship.name;
	} else {
		return miss;
	}
}

var grid = new Grid();
var ship5 = new Ship('5-length', 5, true);
grid.addShip(ship5, 2, 3);
grid.printGrid();
console.log(grid.makeGuess(0, 0));
grid.printGrid();
console.log(grid.makeGuess(2, 3));
grid.printGrid();

